### PR TITLE
Fix jitter incoming message

### PR DESCRIPTION
**Pull request by ihorserba**
Requires also changes in pcm-player:
https://github.com/zelloptt/pcm-player/pull/7

@megamk, @vocoded - could you please take a look

### CODE REVIEW

File: examples/stream.js [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+    }
+    
+    getNextOggPacketStart(onCompleteCb) {
+        // Each Ogg page starts with magic bytes "OggS"
+        // Stream may be corrupted, so find a first valid magic
+        var verifiedBytes = 0;
+        var magic = new TextEncoder("utf-8").encode("OggS");
+        var readbyte = new Uint8Array(1);
+        
+        var readByteCb = function(err, bytesRead, buf) {
```
Comment (megamk) [inline examples/stream.js:37]: `buf` not used

File: examples/stream.js [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+        // |  Version = 1  | Channel Count |           Pre-skip            | 8-11
+        // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        // |                     Input Sample Rate (Hz)                    | 12-15
+        // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        // |   Output Gain (Q7.8 in dB)    | Mapping Family|               | 16-
+        // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               :
+        // :               Optional Channel Mapping Table...               :
+        // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        var version, channels, preskip;
+        var magic = new Uint8Array("OpusHead");
```
Comment (megamk) [inline examples/stream.js:158]: Getting IDE warning here, `Uint8Array` doesn't have a constructor that accepts a string

File: examples/stream.js [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+        
+        for (let i = 0; i < magic.length; i++) {
+            if (magic[i] !== data[i])
+                return false;
```
Comment (megamk) [inline examples/stream.js:162]: we are not using short conditions in js files, could you replace it with 
```js
            if (magic[i] !== data[i]) {
                return false;
            }
```

File: examples/stream.js [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+        // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        // :                   User Comment #0 String...                   :
+        // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        // |                 User Comment #1 String Length                 |
+        // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        var magic = new Uint8Array("OpusTags");
+        
+        for (let i = 0; i < magic.length; i++) {
+            if (magic[i] !== data[i])
+                return false;
```
Comment (megamk) [inline examples/stream.js:201]: same as above, wrap condition with `{}`

File: examples/stream.js [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+        // |0 1 2 3 4 5 6 7| Bit
+        // +-+-+-+-+-+-+-+-+
+        // | config  |s| c |
+        // +-+-+-+-+-+-+-+-+
+        var framesPerPacket;
+        var tocC = (data[0] & 0x03);
+        if (tocC === 0) {
+            framesPerPacket = 1;
+        } else {
+            if (tocC === 1 || tocC === 2) {
```
Comment (megamk) [inline examples/stream.js:218]: can shorten with `elseif`

File: examples/stream.js [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+            [durations[2], new Set([0, 4, 8, 12, 14, 18, 22, 26, 30])],
+            [durations[3], new Set([1, 5, 9, 13, 15, 19, 23, 27, 31])],
+            [durations[4], new Set([2, 6, 10])],
+            [durations[5], new Set([3, 7, 11])]
+        ]);
+        configsMS.forEach(function(item, index) {
+            if (item.has(conf))
+                return [framesPerPacket, durations[index]]
+        });
+        return [framesPerPacket, 20];
```
Comment (megamk) [inline examples/stream.js:241]: this looks like `return [framesPerPacket, durations[index]]` will only return from `forEach` callback and the function itself will always return ` [framesPerPacket, 20];`

File: examples/stream.js [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+        password: password,
+        channel: channel,
+    }));
+    
+    var isAuthorized = false, isChannelAvailable = false;
+    ws.onmessage = function(event) {
+        var json = JSON.parse(event.data);
+        if (json.refresh_token) {
+            isAuthorized = true;
+        } else if (json.command == "on_channel_status" && json.status == "online") {
```
Comment (megamk) [inline examples/stream.js:368]: is it possible to use `===` here?

File: examples/stream.js [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+
+function zelloStartStream(ws, opusStream, onCompleteCb) {
+    var codecHeaderRaw = new Uint8Array(4);
+    codecHeaderRaw[2] = opusStream.framesPerPacket;
+    codecHeaderRaw[3] = opusStream.packetDurationMs;
+    
+    // sampleRate is represented in two bytes in little endian.
+    // https://github.com/zelloptt/zello-channel-api/blob/409378acd06257bcd07e3f89e4fbc885a0cc6663/sdks/js/src/classes/utils.js#L63
+    codecHeaderRaw[0] = parseInt(opusStream.sampleRate & 0xff);
+    codecHeaderRaw[1] = parseInt(opusStream.sampleRate / 0x100) & 0xff;
```
Comment (megamk) [inline examples/stream.js:385]: it's safer to always provide `radix` argument for [`parseInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt) function

File: examples/stream.js [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+    var id = streamId;
+    for (let i = 4; i > 0; i--) {
+        packet[i] = parseInt(id & 0xff);
+        id = parseInt(id / 0x100);
+    }
+        
+    id = packetId;
+    for (let i = 8; i > 4; i--) {
+        packet[i] = parseInt(id & 0xff);
+        id = parseInt(id / 0x100);
```
Comment (megamk) [inline examples/stream.js:426]: same as above, provide `radix` argument for `parseInt`

File: examples/stream.js [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+        return onCompleteCb();
+    }
+    setTimeout(onCompleteCb, sleepDelayMs);
+}
+
+function zelloStreamSendAudio(ws, opusStream, streamId, onCompleteCb) {
+    var startTsMs = getCurrentTimeMs();
+    var timeStreamingMs = 0;
+    var packetId = 0;
+    zelloStreamNextPacket = function() {
```
Comment (megamk) [inline examples/stream.js:447]: missing declaration type here:
```js
const zelloStreamNextPacket = function() {
```

File: examples/stream.js [hunk 11]
```diff
@@ -0,0 +1,547 @@
...
+            let packet = zelloGenerateAudioPacket(data, streamId, packetId);
+            timeStreamingMs += opusStream.packetDurationMs;
+            packetId++;
+            zelloSendAudioPacket(ws, packet, startTsMs, timeStreamingMs, function() {          
+                return zelloStreamNextPacket();
+            });
+        });
+    }
+    zelloStreamNextPacket();
+    ws.onmessage = function(event) {
```
Comment (megamk) [inline examples/stream.js:463]: `event` not used

File: examples/stream.js [hunk 12]
```diff
@@ -0,0 +1,547 @@
...
+}
+
+function zelloStreamReadyCb(opusStream, username, password, token, channel) {
+    var ws = new WebSocket("wss://zello.io/ws");
+    ws.onclose = function(event) {
+        zelloSocket = null;
+        process.exit();
+    };
+                    
+    ws.onopen = function(event) {
```
Comment (megamk) [inline examples/stream.js:481]: `event` not used

File: examples/stream.js [hunk 13]
```diff
@@ -0,0 +1,547 @@
...
+        zelloSocket = ws;
+        
+        zelloAuthorize(ws, opusStream, username, password, token, channel, function(success) {
+            if (!success) {
+                console.log("Failed to authorize");
```
Comment (megamk) [inline examples/stream.js:486]: `console.error()` or `console.trace()`

File: examples/stream.js [hunk 14]
```diff
@@ -0,0 +1,547 @@
...
+                ws.close();
+            }
+            zelloStartStream(ws, opusStream, function(streamId) {
+                if (!streamId) {
+                    console.log("Failed to start Zello stream");
```
Comment (megamk) [inline examples/stream.js:491]: `console.error()` or `console.trace()`

File: examples/stream.js [hunk 15]
```diff
@@ -0,0 +1,547 @@
...
+                    ws.close();
+                }
+                zelloStreamId = streamId;
+                zelloStreamSendAudio(ws, opusStream, streamId, function(success) {
+                    if (!success) {
+                        console.log("Failed to stream audio");
```
Comment (megamk) [inline examples/stream.js:497]: `console.error()` or `console.trace()`

**Comment by ihorserba** [comment]
> Did not test. Code-style comments mostly with one possible bug

Changed according the comments.

**Comment by ihorserba** [comment]
Moved away implementation of the API example clients.
It lives here: https://github.com/zelloptt/zello-channel-api/pull/102

File: sdks/js/src/classes/incomingMessage.js [hunk 16]
```diff
@@ -26,7 +26,7 @@ class IncomingMessage extends Emitter {
         encoding: '32bitFloat',
         channels: 1,
         sampleRate: IncomingMessage.detectSampleRate(this.codecDetails.rate),
-        flushingTime: 300
+        flushingTime: this.codecDetails.frameSize * 10,
```
Comment (adambailey-) [inline sdks/js/src/classes/incomingMessage.js:29]: Is there a significance to 10? If so, can you please comment?

Comment (ihorserba) [reply sdks/js/src/classes/incomingMessage.js:29]: > Is there a significance to 10? If so, can you please comment?

The idea was to flush the PCM buffer every **N** audio frames to have the same audio duration on each iteration.

I took 10 as it gives us 200ms on most common 20ms frame duration - close enough to the original value of 300ms.

The idea is just theoretical - if you think this commit won't improve the audio quality - lets discard it.

Comment (adambailey-) [reply sdks/js/src/classes/incomingMessage.js:29]: Yeah, I think for the current value of 20ms frame sizes this change is fine, but on other Zello apps 60ms frames are normal, and I believe opus even allows 120ms frames. 600-1200 ms flushes are quite a bit larger than before

Comment (ihorserba) [reply sdks/js/src/classes/incomingMessage.js]: > Yeah, I think for the current value of 20ms frame sizes this change is fine, but on other Zello apps 60ms frames are normal, and I believe opus even allows 120ms frames. 600-1200 ms flushes are quite a bit larger than before

Agree. I've changed the flush period to 240 as the Least Common Multiple of the possible durations - https://github.com/zelloptt/zello-channel-api/pull/112/commits/c02713df55fecda1d8b10ba2ba314a658d6ed108

**Review by vocoded — APPROVED** [review]

**Pull request merged**